### PR TITLE
Azure Monitor: Change response to be dataframes

### PR DESCRIPTION
**Pull request by kylebrandt**

### CODE REVIEW

File: pkg/tsdb/azuremonitor/azuremonitor-datasource.go [hunk 1]
```diff
@@ -260,25 +260,33 @@ func (e *AzureMonitorDatasource) unmarshalResponse(res *http.Response) (AzureMon
...
-		metricName := formatAzureMonitorLegendKey(query.Alias, query.UrlComponents["resourceName"], data.Value[0].Name.LocalizedValue, metadataName, metadataValue, data.Namespace, data.Value[0].ID)
+		metricName := formatAzureMonitorLegendKey(query.Alias, query.UrlComponents["resourceName"], amr.Value[0].Name.LocalizedValue, metadataName, metadataValue, amr.Namespace, amr.Value[0].ID)
 
-		for _, point := range series.Data {
+		frame := data.NewFrameOfFieldTypes(query.Alias, len(series.Data), data.FieldTypeTime, data.FieldTypeFloat64)
+		frame.Fields[1].Name = metricName
+		frame.Fields[1].SetConfig(&data.FieldConfig{
+			Unit: amr.Value[0].Unit,
+		})
+
```
Comment (kylebrandt) [inline pkg/tsdb/azuremonitor/azuremonitor-datasource.go:284]: Need to set RefId here (at least for now). If we stop encoding the frames too early in the process, then easy to add it when marshaling the frames (like we do when converting to protobuf with the SDK).

**Comment by kylebrandt** [comment]
Not sure what to change at https://github.com/grafana/grafana/blob/fcb9dcb0b8ab15ff9a19c1c59bf93423fcb5e4ee/public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts#L120-L137 ... to many `any` for me :-)

But the base64 outputed loaded in Grafana dev datasource looks correctish.

**Comment by ryantxu** [comment]
Ideally we can have the frontend extend [DataSourceWithBackend.ts](https://github.com/grafana/grafana/blob/master/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L45) -- the `map` part of the query would live in applyTemplateVariables

should the filter part happen on the frontend?  or in the backend?

**Comment by ryantxu** [comment]
@kylebrandt -- totally untested, but at least conceptually it should work ;)

The tests are failing because the datasource is really three things glued together, and I don't yet have them all mocked.  If the raw results are working as expected I'll spend some more time trying to get it to work as expected

**Comment by kylebrandt** [comment]
Ryan is going to update frontend tests a little later, but backend should be good for review.

**Comment by kylebrandt** [comment]
@ryantxu I think this seems to break "text edit mode" in application insights:

"If you change the service type to "Application Insights", the menu icon to the right adds another option, "Toggle Edit Mode". Once clicked, the query edit mode changes to give you a full text area in which to write log analytics queries. (This is identical to how the InfluxDB datasource lets you write raw queries.)"

In 7.0 this changes the query editor to have a little text entry line:

image

But this seems to be a no-op on this.

**Comment by ryantxu** [comment]
I am unable to get azure to do anything and do not see anything happen when I click the edit icon :(

If you type in a query to the raw text box, what is the saved panel JSON object?

**Comment by kylebrandt** [comment]
Query:

image


Panel JSON:

```
 "targets": [
        {
          "appInsights": {
            "aggOptions": [
              "sum"
            ],
            "aggregation": "sum",
            "dimension": "request/source",
            "dimensions": [
              "request/source",
              "request/name",
              "request/urlHost",
              "request/urlPath",
              "request/success",
              "request/resultCode",
              "request/performanceBucket",
              "operation/name",
              "operation/synthetic",
              "operation/syntheticSource",
              "user/authenticated",
              "application/version",
              "client/type",
              "client/model",
              "client/os",
              "client/city",
              "client/stateOrProvince",
              "client/countryOrRegion",
              "client/browser",
              "cloud/roleName",
              "cloud/roleInstance",
              "customDimensions/_MS.ProcessedByMetricExtractors"
            ],
            "metricName": "requests/count",
            "rawQuery": true,
            "rawQueryString": "Select Thanks From Ryan",
            "timeColumn": "timestamp",
            "timeGrain": "auto",
            "valueColumn": ""
          },
          "azureLogAnalytics": {
            "query": "//change this example to create your own time series query\n<table name>                                                              //the table to query (e.g. Usage, Heartbeat, Perf)\n| where $__timeFilter(TimeGenerated)                                      //this is a macro used to show the full chart’s time range, choose the datetime column here\n| summarize count() by <group by column>, bin(TimeGenerated, $__interval) //change “group by column” to a column in your table, such as “Computer”. The $__interval macro is used to auto-select the time grain. Can also use 1h, 5m etc.\n| order by TimeGenerated asc",
            "resultFormat": "time_series",
            "workspace": "REDACTED"
          },
          "azureMonitor": {
            "aggOptions": [],
            "dimensionFilter": "*",
            "metricDefinition": "select",
            "metricName": "select",
            "metricNamespace": "select",
            "resourceGroup": "select",
            "resourceName": "select",
            "timeGrain": "auto",
            "timeGrains": [],
            "top": "10"
          },
          "queryType": "Application Insights",
          "refId": "A",
          "subscription": "REDACTED"
        }
```

File: pkg/tsdb/azuremonitor/azuremonitor-datasource.go [hunk 2]
```diff
@@ -292,15 +299,17 @@ func (e *AzureMonitorDatasource) parseResponse(queryRes *tsdb.QueryResult, data
 			default:
 				value = point.Count
 			}
-			points = append(points, tsdb.NewTimePoint(null.FloatFrom(value), float64((point.TimeStamp).Unix())*1000))
+
+			frame.SetRow(i, point.TimeStamp, value)
```
Comment (marefr) [inline pkg/tsdb/azuremonitor/azuremonitor-datasource.go:303]: Shouldn't timestamp be in epoch ms? Looks like old code converts from eppch seconds to ms?

Comment (marefr) [reply pkg/tsdb/azuremonitor/azuremonitor-datasource.go:303]: Or maybe I'm wrong? Confused by the `tsdb.NewTimePoint`...

**Review by marefr — APPROVED** [review]
LGTM besides the comment I added. Tested that alerting with azure monitor works (basic test).

File: pkg/tsdb/azuremonitor/azuremonitor-datasource.go (hunk 2 above)
Comment (kylebrandt) [reply pkg/tsdb/azuremonitor/azuremonitor-datasource.go:303]: Oh sorry I missed this comment. Time in dataframes are now time.Time objects.

Comment (kylebrandt) [reply pkg/tsdb/azuremonitor/azuremonitor-datasource.go:303]: That being said, when turned into Arrow they do become Nanosecond epochs:

https://github.com/grafana/grafana-plugin-sdk-go/blob/6561b72621d71a8efa5be36f2469e1c0eb003ada/data/arrow_column_builders.go#L393-L401

**Pull request merged**